packages/regenerator-transform/src/visit.js

```diff
@@ -177,32 +184,118 @@ function getOuterFnExpr(funPath) {
     node.id = funPath.scope.parent.generateUidIdentifier("callee");
   }
 
+  let outerFnExpr = node.id
+
   if (node.generator && // Non-generator functions don't need to be marked.
-      t.isFunctionDeclaration(node)) {
-    let pp = funPath.findParent(function (path) {
+    t.isFunctionDeclaration(node)) {
+    let pp = funPath.findParent(function(path) {
       return path.isProgram() || path.isBlockStatement();
     });
 
     if (!pp) {
-      return node.id;
+      return funPath;
     }
+    const parameters = funPath.node.params.map(function(param) {
+      return t.cloneDeep(param);
+    })
+
+    // turn our FunctionDeclaration into a FunctionExpression with a different
+    // name - our our eventual outerFnExpr needs to be identical to the original
+    // FunctionDeclaration name so that external calls work correctly, and we
+    // can't also use that name in our FunctionExpression because it would shadow
+    // the outerFnExpr in our wrapping scope
+    const functionExpression = t.functionExpression(
+      funPath.scope.generateUidIdentifierBasedOnNode(node.id),
+      parameters,
+      t.cloneDeep(funPath.node.body),
+      funPath.node.generator,
+      funPath.node.async
+    );
 
-    let markDecl = getRuntimeMarkDecl(pp);
-    let markedArray = markDecl.declarations[0].id;
-    let funDeclIdArray = markDecl.declarations[0].init.callee.object;
-    t.assertArrayExpression(funDeclIdArray);
+    // in spite of the above, we want the resulting function's name property
+    // property to be the same as the original so we wrap the FunctionExpression
+    // with:
+    //   Object.defineProperty(functionExpression, 'name', { value: outerFnExpr.name }
+    // this way introspection / prototype chains work the way we'd expect
```

Comment:
understanding all corner cases here is a little bit tricky and I have limited time now, but from what I understand about the issue:

- we cannot have the same identifier in the outer scope and in the inner scope because it would always act like the [safe example in the tests](https://github.com/facebook/regenerator/blob/9b6ebc0532ff373e06e3421b26800ac8979913d3/test/tests.es6.js#L1604-L1611)
- this leads me to the conclusion that the simplest answer for this is to just preserve transforming to the function declarations (like in the original transform) and not to function expression. What was the logic behind this change? 

It seems that the easiest would be to generate this:
```js
var toBeRebound$$ = /*#__PURE__*/regeneratorRuntime.mark(toBeRebound);
function toBeRebound() {
  var originalValue;
  return regeneratorRuntime.wrap(function toBeRebound$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          originalValue = toBeRebound;
          _context.next = 3;
          return toBeRebound;

        case 3:
          assert.strictEqual(rebindTo(42), originalValue);
          _context.next = 6;
          return toBeRebound;

        case 6:
          assert.strictEqual(rebindTo("asdf"), 42);
          _context.next = 9;
          return toBeRebound;

        case 9:
        case "end":
          return _context.stop();
      }
    }
  }, toBeRebound$$, this);
};
function rebindTo(value) {
  var oldValue = toBeRebound;
  toBeRebound = value;
  return oldValue;
}
```
This would:
- keep the `name` property untouched and make `Object.defineProperty` calls obsolete.
- keep proper identifiers in the proper scopes
- keep original reference (before rebinding) in scope so it can be accessed by the `regeneratorRuntime.wrap` call (which was previously accessed with `_marked[0]`)